[MM-48020] work template preview page fade illustration animation

mhd-sln opened the pull request:
#### Summary
Add fade animation for illustration on the preview page in the work template

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-48020

#### Video
~~https://www.loom.com/share/23c54e8887034ccf91f697cc0dbed536~~
latest: https://www.loom.com/share/4886c1b6cd794f92ad3c24b46ced94a5

#### Release Note
```release-note
NONE
```

pvev commented on components/worktemplates/utils.ts:
```diff
@@ -3,31 +3,31 @@
     const channels = template.content.filter((c) => c.channel).map((c) => c.channel!);
     if (channels.length) {
-        return channels[0].illustration;
+        return ['channels', channels[0].illustration];
```
nit: good practice to avoid magic strings, use an enum to hold these values is a better practice.

pvev reviewed (approved):
LGTM! thanks @mhd-sln :shipit:

JulienTant commented:
I did not check the code yet, just the UI - I think the effect is supposed to be that the first picture fade-out and the new one fade-in. Tagging @annelieseklein on this PR to ensure I'm not saying something wrong.

mhd-sln commented:
Looking at this again. You might be right @JulienTant.

julmondragon commented on components/worktemplates/components/modal_body_with_illustration.tsx:
```diff
@@ -1,26 +1,42 @@
 export interface ModalBodyWithIllustrationProps {
     className?: string;
-    illustration: string;
+    illustration: [string, string];
```
Using an object would help making more clear what's the expected data here rather than indexes `{type: string; src: string}`. Or even having individual props?

JulienTant commented on components/worktemplates/components/modal_body.tsx:
```diff
@@ -0,0 +1,35 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
```
I think this file is not needed at all. It's only used for this page preview page so we might as well put everything in the page.

JulienTant commented on components/worktemplates/components/preview.tsx:
```diff
@@ -20,10 +23,54 @@ export interface PreviewProps {
+interface IllustrationAnimations {
+    prior: {
+        mounted: boolean;
+        type?: TemplateType;
```
We only want to animate when the illustration change, I believe the type is unnecessary.

JulienTant commented on components/worktemplates/components/preview.tsx:
```diff
@@ -59,7 +106,14 @@ const Preview = ({template, ...props}: PreviewProps) => {
                     message={template.description.channel.message}
                     items={channels}
-                    onUpdateIllustration={setCurrentIllustration}
+                    onUpdateIllustration={(id: string, illustration: string) => setIllustrationDetails({
+                        prior: {...illustrationDetails.current},
+                        current: {
+                            mounted: false,
+                            type: TemplateType.CHANNELS,
+                            illustration,
+                        },
+                    })}
```
Also we have this bit of code in a lot of places, it might be worth it to make a method to improve readability.

```suggestion
                    onUpdateIllustration={handleUpdateIllustration}
```

JulienTant reviewed (commented):
A few remarks/questions here and there but this is looking great!

julmondragon reviewed (approved):
Looks good @mhd-sln !
Just some nits, and a couple of questions:

![Nov-09-2022 09-41-38](https://user-images.githubusercontent.com/79058848/200864005-6b3dc077-cfc9-43c5-be18-7b88607dc7eb.gif)

- Is the fading animation expected to fire when closing/opening the current type?
- When closing the current type, should it remain as selected (blue border)?

neallred commented on components/worktemplates/components/preview.tsx:
```diff
@@ -147,6 +288,60 @@ const StyledPreview = styled(Preview)`
+      opacity: 1;
+    }
+
+    .current-illustration-exit {
```
The opacity rules could be combined like this for more brevity:

```css
.prior-illustration-enter,
.prior-illustration-enter-done,
.prior-illustration-exit-done {
  opacity: 0;
}

.current-illustration-enter,
.current-illustration-exit,
.current-illustration-exit-done {
  opacity: 0;
}
```

mhd-sln commented:
> Looks good @mhd-sln ! Just some nits, and a couple of questions:
> 
> ![Nov-09-2022 09-41-38](https://user-images.githubusercontent.com/79058848/200864005-6b3dc077-cfc9-43c5-be18-7b88607dc7eb.gif) [ ![Nov-09-2022 09-41-38](https://user-images.githubusercontent.com/79058848/200864005-6b3dc077-cfc9-43c5-be18-7b88607dc7eb.gif) ](https://user-images.githubusercontent.com/79058848/200864005-6b3dc077-cfc9-43c5-be18-7b88607dc7eb.gif) [ ](https://user-images.githubusercontent.com/79058848/200864005-6b3dc077-cfc9-43c5-be18-7b88607dc7eb.gif)
> 
> * Is the fading animation expected to fire when closing/opening the current type?
> * When closing the current type, should it remain as selected (blue border)?

Great questions, @julmondragon. 
Regarding the fading animation for closing/opening the current type, I think it shouldn't fire. 
For the second question, I am not sure about it.
Tagging @annelieseklein here to have her input on these.

mhd-sln commented:
Thank you folks for the insightful comments. 👍🏻 
I have implemented the suggested improvement can you please go through them once more ? @JulienTant @neallred @julmondragon. 
I have also implemented the lack of fade animation when clicking on the same type @julmondragon

JulienTant reviewed (approved):
Awesome! Thank you @mhd-sln.

julmondragon commented on components/worktemplates/components/preview.tsx:
```diff
@@ -20,10 +22,64 @@ export interface PreviewProps {
+    exit: 200,
+};
+
 const Preview = ({template, ...props}: PreviewProps) => {
```
nit: `{template, className}`

julmondragon reviewed (approved):
LGTM!

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/261171

stevemudie reviewed (approved):
This looks great @mhd-sln!

mhd-sln commented:
Discussed this with @JulienTant and we thought that it might be a good idea to merge this and https://github.com/mattermost/mattermost-webapp/pull/11523 into https://github.com/mattermost/mattermost-webapp/pull/11475.
We would be revisiting the code from UX review perspective later as well once the Work Template Modal would be feature ready in the future. @annelieseklein
